### PR TITLE
DG material coverage check

### LINKED ISSUE #9482: getMaterialProperty in DGKernel is not checked for integrity
### Description of the enhancement or error report
Looks like `DGKernel::getMaterialProperty` is not mark the consumer of the material property, thus the problem integrity check is not covering the case where `DGKernel` is trying to use a un-declared property.

### Rationale for the enhancement or information for reproducing the error
DG kernels may not properly set up.

### Identified impact
(i.e. Internal object changes, limited interface changes, public API change, or a list of specific applications impacted)
Input robustness.

### PR DESCRIPTION
**Pull request by YaqiWang**
Refer to #9482.

### CODE REVIEW

**Comment by permcody** [comment]
@lindsayad - I see that this check breaks one of your tests. Could you confirm that there is indeed a problem?

**Comment by YaqiWang** [comment]
I cannot see the test?

**Comment by permcody** [comment]
@YaqiWang - Log in, if you still can't see it, let us know.

File: framework/src/dgkernels/DGKernel.C [hunk 1]
```diff
@@ -65,7 +65,7 @@ DGKernel::DGKernel(const InputParameters & parameters)
     FunctionInterface(this),
     UserObjectInterface(this),
     NeighborCoupleableMooseVariableDependencyIntermediateInterface(this, false, false),
-    TwoMaterialPropertyInterface(this),
+    TwoMaterialPropertyInterface(this, boundaryRestricted() ? std::set<SubdomainID>() : blockIDs(), boundaryIDs()),
```
Comment (YaqiWang) [inline framework/src/dgkernels/DGKernel.C]: I checked the failing test in Zapdos. Nothing wrong there. For interface kernels, the coverage check is much more complicated. We need to call `markMatPropRequested` differently in `getMaterialProperty` and `getNeighborMaterialProperty` based on the subdomains attaching the internal boundary. The material property could be provided by block or boundary materials on the master side, which makes things more complicated. @permcody has an idea in his mind that we the coverage check in front by looping though elements and sides and deprecated the current checking method.

Thus, I have `boundaryRestricted() ? std::set<SubdomainID>() : blockIDs(), boundaryIDs()` here as a heck to enable the check for `DGKernel` and `InternalSideUserObject` but skip the check for `InterfaceKernel`. I think this is worthwhile to be pushed into framework, but need to have others' opinion.

File: framework/src/dgkernels/DGKernel.C [hunk 2]
```diff
@@ -65,7 +65,8 @@ DGKernel::DGKernel(const InputParameters & parameters)
     FunctionInterface(this),
     UserObjectInterface(this),
     NeighborCoupleableMooseVariableDependencyIntermediateInterface(this, false, false),
-    TwoMaterialPropertyInterface(this),
+    TwoMaterialPropertyInterface(
+        this, boundaryRestricted() ? std::set<SubdomainID>() : blockIDs(), boundaryIDs()),
```
Comment (YaqiWang) [inline framework/src/dgkernels/DGKernel.C]: I checked the failing test in Zapdos. Nothing wrong there. For interface kernels, the coverage check is much more complicated. We need to call markMatPropRequested differently in `getMaterialProperty` and `getNeighborMaterialProperty` based on the subdomains attaching the internal boundary. The material property could be provided by block or boundary materials on the master side, which makes things more complicated. @permcody has an idea in his mind that we the coverage check in front by looping though elements and sides and deprecated the current checking method.

Thus, I have `boundaryRestricted() ? std::set<SubdomainID>() : blockIDs(), boundaryIDs()` here as a heck to enable the check for `DGKernel` and `InternalSideUserObject` but skip the check for InterfaceKernel. I think this is worthwhile to be pushed into framework, but need to have others' opinion.

Comment (permcody) [inline framework/src/dgkernels/DGKernel.C]: At the very least you need to document the hack in the source code so we can find it later and remove it.

**Comment by YaqiWang** [comment]
@permcody This fix is good now. I do not think we have complete boundary material coverage test, but that can be done in the two places I marked with `FIXME` in this PR.

**Comment by lindsayad** [comment]
There must have been a force push here because it looks like all Zapdos tests are working...?

**Comment by YaqiWang** [comment]
@lindsayad Yes.

**Comment by YaqiWang** [comment]
@permcody, I will fix the failing test caused by a mesh modifier not supporting distributed mesh. Other than this, this PR is good for review.

File: framework/src/materials/MaterialPropertyInterface.C [hunk 3]
```diff
@@ -50,6 +52,8 @@ MaterialPropertyInterface::MaterialPropertyInterface(const MooseObject * moose_o
     _mi_tid(_mi_params.get<THREAD_ID>("_tid")),
     _stateful_allowed(true),
     _get_material_property_called(false),
+    _block_restricted(block_ids.find(Moose::ANY_BLOCK_ID) == block_ids.end()),
```
Comment (permcody) [inline framework/src/materials/MaterialPropertyInterface.C]: This doesn't seem like the correct design. MaterialPropertyInterface shouldn't have these block/boundary restrictable Booleans as members. It's just an interface class. Can they not be moved to the TwoMaterialPropertyInterface class which is intended to understand block/boundary restriction? At the very least, you should be obtaining this information from the Block/Boundary restrictable objects directly, not recoding it here.

Comment (permcody) [inline framework/src/materials/MaterialPropertyInterface.C]: Yeah, looking through this PR, I don't see where these Booleans are used in this class. Only the derived class where you have access to the other objects for this information.

**Review by permcody — CHANGES_REQUESTED** [review]

**Comment by permcody** [comment]
This PR is just about ready to hit our 30 day inactivity limit.

**Comment by YaqiWang** [comment]
Although the check is necessary, the fix proposed here may not be right. I will let it die and leave the issue open.

**Comment by YaqiWang** [comment]
BTW, those three tests here have value and can be used with the right fix.

**Comment by permcody** [comment]
@aeslaughter - Just pinging you on this. I don't think @YaqiWang is going to work on this further and I'm not sure it's critical for anybody to jump on at the moment. Probably will close for now.

**Comment by permcody** [comment]
@aeslaughter - Any interest in fixing this? Perhaps your other change will help here?

**Comment by aeslaughter** [comment]
This line should fix this: https://github.com/idaholab/moose/blob/devel/framework/src/dgkernels/DGKernel.C#L81

**Comment by YaqiWang** [comment]
Look like the two tests I added in my last commit 18e60a2 do not error out still. Do I miss something? The DG test in 5c033e6 properly errors out.

**Comment by aeslaughter** [comment]
@YaqiWang You are correct, your tests do not throw an error.

You have created a "block" material, but are using them in a boundary restricted object (BC). This is a separate issue I will need to investigate to understand better. I am not sure what we are doing in this case.

**Pull request closed without merging**